File: 01_Arrays_&_Hashing/01_Contains_Duplicate/0217-contains-duplicate.py
"""
Problem: LeetCode 217 - Contains Duplicate

Key Idea:
To check if there are any duplicates in the given list, we can use a hash set (set in Python) to store the unique elements as we traverse the list. For each element, we check if it is already present in the set. If it is, then we have found a duplicate, and we return True. If we traverse the entire list without finding any duplicates, we return False.

Time Complexity:
The time complexity of this approach is O(n), where n is the size of the input list. In the worst case, we might need to traverse the entire list once to find a duplicate.

Space Complexity:
The space complexity is O(n), as the hash set can potentially store all elements of the input list if they are all distinct.
"""


class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        hashset = set()

        for n in nums:
            if n in hashset:
                return True
            hashset.add(n)
        return False

File: 01_Arrays_&_Hashing/01_Contains_Duplicate/0217-contains-duplicate.rs
/*
Problem: LeetCode 217 - Contains Duplicate

Key Idea:
The key idea is to use a hashset to keep track of the numbers we have seen so far. If we encounter a number that is already in the hashset, we know there is a duplicate.

Approach:
1. Create an empty hashset to store the numbers and their occurrences.
2. Iterate through the input array:
   - For each number `num`, check if it exists in the hashset. If yes, return `true` as a duplicate is found.
   - Otherwise, insert `num` into the hashset.
3. If no duplicate is found during the iteration, return `false`.

Time Complexity:
O(n), where n is the number of elements in the input array. We iterate through the array once and each hashset lookup takes O(1) time on average.

Space Complexity:
O(n), as the worst case is when all elements in the array are distinct and we store all of them in the hashset.
*/

use std::collections::HashSet;

impl Solution {
    pub fn contains_duplicate(nums: Vec<i32>) -> bool {
        let mut seen_nums = HashSet::new();

        for num in nums {
            if seen_nums.contains(&num) {
                return true;
            }
            seen_nums.insert(num);
        }

        false
    }
}
